lib/rust/flexer/src/automata/pattern.rs

```diff
@@ -132,24 +132,72 @@ impl Pattern {
 
 impl BitOr<Pattern> for Pattern {
     type Output = Pattern;
-    fn bitor(self, rhs: Pattern) -> Self::Output {
+    fn bitor(self, rhs:Pattern) -> Self::Output {
         match (self, rhs) {
-            (Or(mut lhs), Or(    rhs)) => {lhs.extend(rhs) ; Or(lhs)},
-            (Or(mut lhs), rhs        ) => {lhs.push(rhs)   ; Or(lhs)},
-            (lhs        , Or(mut rhs)) => {rhs.push(lhs)   ; Or(rhs)},
+            (Or(mut lhs), Or(    rhs)) => {lhs.extend(rhs)   ; Or(lhs)},
+            (Or(mut lhs), rhs        ) => {lhs.push(rhs)     ; Or(lhs)},
+            (lhs        , Or(mut rhs)) => {rhs.insert(0,lhs) ; Or(rhs)},
             (lhs        , rhs        ) => Or(vec![lhs,rhs]),
         }
     }
 }
 
+impl BitOr<&Pattern> for &Pattern {
+    type Output = Pattern;
+
+    fn bitor(self, rhs:&Pattern) -> Self::Output {
+        self.clone() | rhs.clone()
+    }
+}
+
+impl BitOr<&Pattern> for Pattern {
+    type Output = Pattern;
+
+    fn bitor(self, rhs:&Pattern) -> Self::Output {
+        self | rhs.clone()
+    }
+}
+
+impl BitOr<Pattern> for &Pattern {
+    type Output = Pattern;
+
+    fn bitor(self, rhs:Pattern) -> Self::Output {
+        self.clone() | rhs
+    }
+}
+
 impl Shr<Pattern> for Pattern {
     type Output = Pattern;
-    fn shr(self, rhs: Pattern) -> Self::Output {
+    fn shr(self, rhs:Pattern) -> Self::Output {
         match (self, rhs) {
-            (Seq(mut lhs), Seq(rhs))     => {lhs.extend(rhs) ; Seq(lhs)},
-            (Seq(mut lhs), rhs         ) => {lhs.push(rhs)   ; Seq(lhs)},
-            (lhs         , Seq(mut rhs)) => {rhs.push(lhs)   ; Seq(rhs)},
+            (Seq(mut lhs), Seq(rhs)    ) => {lhs.extend(rhs)   ; Seq(lhs)},
+            (Seq(mut lhs), rhs         ) => {lhs.push(rhs)     ; Seq(lhs)},
+            (lhs         , Seq(mut rhs)) => {rhs.insert(0,lhs) ; Seq(rhs)},
             (lhs         , rhs         ) => Seq(vec![lhs, rhs]),
         }
     }
 }
+
+impl Shr<&Pattern> for &Pattern {
+    type Output = Pattern;
+
+    fn shr(self, rhs:&Pattern) -> Self::Output {
+        self.clone() >> rhs.clone()
+    }
+}
+
+impl Shr<&Pattern> for Pattern {
+    type Output = Pattern;
+
+    fn shr(self, rhs:&Pattern) -> Self::Output {
+        self >> rhs.clone()
+    }
+}
+
+impl Shr<Pattern> for &Pattern {
+    type Output = Pattern;
+
+    fn shr(self, rhs:Pattern) -> Self::Output {
+        self.clone() >> rhs
+    }
+}
```

Comment:
this is code repetition for Shr and BitOr - create macro rules here and use it for both things instead.